echo-pubsub/src/main/java/com/netflix/spinnaker/echo/pubsub/amazon/SQSSubscriber.java

```diff
@@ -0,0 +1,255 @@
+import com.netflix.spinnaker.echo.pubsub.utils.NodeIdentity;
+import com.netflix.spinnaker.kork.artifacts.model.Artifact;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.*;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+/**
+ * One subscriber for each subscription.
+ * The subscriber makes sure the SQS queue is created, subscribes to the SNS topic,
+ * polls the queue for messages, and removes them once processed.
+ */
+public class SQSSubscriber implements Runnable, PubsubSubscriber {
+
+  private static final Logger log = LoggerFactory.getLogger(SQSSubscriber.class);
+
+  private static final int AWS_MAX_NUMBER_OF_MESSAGES = 10;
+  static private final PubsubSystem pubsubSystem = PubsubSystem.AMAZON;
+
+  private ObjectMapper objectMapper;
+  private AmazonSNS amazonSNS;
+  private AmazonSQS amazonSQS;
+
+  private AmazonPubsubProperties.AmazonPubsubSubscription subscription;
+
+  private PubsubMessageHandler pubsubMessageHandler;
+  private MessageArtifactTranslator messageArtifactTranslator;
+
+  private NodeIdentity identity = new NodeIdentity();
+
+  Registry registry;
+
+  private final ARN queueARN;
+  private final ARN topicARN;
+
+  private String queueId = null;
+
+
+  public SQSSubscriber(ObjectMapper objectMapper,
+                       AmazonPubsubProperties.AmazonPubsubSubscription subscription,
+                       PubsubMessageHandler pubsubMessageHandler,
+                       AmazonSNS amazonSNS,
+                       AmazonSQS amazonSQS,
+                       Registry registry) {
+    this.objectMapper = objectMapper;
+    this.subscription = subscription;
+    this.pubsubMessageHandler = pubsubMessageHandler;
+    this.amazonSNS = amazonSNS;
+    this.amazonSQS = amazonSQS;
+    this.registry = registry;
+
+    this.messageArtifactTranslator = new MessageArtifactTranslator(subscription.readTemplatePath());
+    this.queueARN = new ARN(subscription.getQueueARN());
+    this.topicARN = new ARN(subscription.getTopicARN());
+  }
+
+  public String getWorkerName() {
+    return queueARN.arn + "/" + SQSSubscriber.class.getSimpleName();
+  }
+
+  @Override
+  public PubsubSystem pubsubSystem() {
+    return pubsubSystem;
+  }
+
+  @Override
+  public String subscriptionName() {
+    return subscription.getName();
+  }
+
+  @Override
+  public String getName() {
+    return getWorkerName();
+  }
+
+  @Override
+  public void run() {
+    log.info("Starting " + getWorkerName());
+
+    while (true) {
+      try {
+        listenForMessages();
+      } catch (Throwable e) {
+        log.error("Unexpected error running " + getWorkerName() + ", restarting", e);
+      }
+    }
+  }
+
+  private void listenForMessages() {
+    this.queueId = ensureQueueExists(
+      amazonSQS, queueARN, topicARN, subscription.getSqsMessageRetentionPeriodSeconds()
+    );
+    subscribeToTopic(amazonSNS, topicARN, queueARN);
+
+    while (true) {
```

Comment:
I think I initially envisioned the pub/sub polling parts/agent looking similar to [Igor's polling parts](https://github.com/spinnaker/igor/blob/279f2f951747acb19eea2d4ae60896267dec9e02/igor-web/src/main/groovy/com/netflix/spinnaker/igor/jenkins/JenkinsBuildMonitor.groovy). Looking back at the fact that Google pub/sub is async and that SQS has its own polling flavor, it seems like that pattern is not a great fit. I still think the interface for integrating to pub/sub systems should be more uniform (i.e. if we add 5 more integrations where there is no standard interface, the code becomes hard to maintain). I think I'm going to take another look at this when I get around to implementing the Kafka integration since that will also be polling to make it more uniform. Thoughts?

Reply:
We use SQS and SNS for some bits in `clouddriver`, and I took most of this logic/flow from https://github.com/spinnaker/clouddriver/blob/master/clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/lifecycle/InstanceTerminationLifecycleWorker.java. This is quite well tested and was easy to reason about, which is why I followed that pattern.

I agree that it would be great to have a better definition for polling implementations since aws isn't async. If you have suggestions now I'm super happy to refactor. I did struggle a bit to get aws to conform to the async model, but overall I think it's not too bad.